client/src/util/check-permission.ts

```diff
@@ -1,18 +1,48 @@
-import type { AuthContextType } from '../modules/auth/store';
 import {
   InstancePermission,
   ChapterPermission,
 } from '../../../common/permissions';
+import { AuthContextType } from '../modules/auth/store';
+
+type User = NonNullable<AuthContextType['user']>;
+
+function isAllowedByChapterRole(
+  user: User,
+  requiredPermission: string,
+  options: { chapterId: number },
+) {
+  const chapterRole = user.user_chapters.find(
+    ({ chapter_id }) => chapter_id === options.chapterId,
+  )?.chapter_role;
+  return chapterRole
+    ? chapterRole.chapter_role_permissions.some(
+        ({ chapter_permission: { name } }) => name === requiredPermission,
+      )
+    : false;
+}
+
+function isAllowedByInstanceRole(user: User, requiredPermission: string) {
+  return user.instance_role.instance_role_permissions.some(
+    ({ instance_permission: { name } }) => name === requiredPermission,
+  );
+}
+
+function isBannedFromChapter(user: User, options: { chapterId: number }) {
+  return user.user_bans.some(
+    ({ chapter_id }) => chapter_id === options.chapterId,
+  );
+}
```

Comment:
I'll dig into it again. I tried to extract some parts already, but I got tangled into some `user` type differences.

Reply:
Righto. It might be that we should be returning everything the auth checker uses as part of the me query. Then the type issues will go away.

It's maybe not super efficient, but it wouldn't require additional db queries.

Reply:
> I'm thinking about extracting part or all of the server side checking and reuse it here as well.

I only just saw this comment :cold_sweat: At least we were thinking along the same lines!